Update KEP-77 with additional details and focus on batchv1/Job (#5648)
kubernetes-sigs/kueue: keps/77-dynamically-sized-jobs/README.md

#### Creation
The creation of WorkloadSlices in Kueue is initiated when a Kueue Job with slicing enabled, via both annotation and feature gate, is admitted for execution. 
Upon admission, the `jobframework.JobReconciler` evaluates the Job’s PodSets and resource requirements, then creates a new or additional (slice) Workload object 
that represents a partition of the original workload. Each slice contains a `PodSet.Count` identical to the Job’s definition, along with scheduling metadata needed for handling preemption of the old workload (slice).

At most two active workload slices can exist for any single Job in Kueue: the current admitted slice and a newly created slice during a scaling operation. 
This ensures a controlled transition between slices while maintaining consistency and avoiding resource contention and race conditions.

To facilitate precise resource scheduling and effective workload preemption, every new job configured with slice enablement will be appropriately annotated using `EnabledAnnotationKey/Value` mentioned above.
Furthermore, each individual workload slice will record the identity of any preempting slice it encounters.

```golang

const (

> 	// WorkloadPreemptibleSliceNameKey is the annotation key used to capture an "old" workload slice
> 	// that will be preempted by the "new" workload slice, i.e., this annotates a "new" workload slice.

Reviewer: ```suggestion
	// WorkloadPreemptibleSliceNameKey is the annotation key used to capture an "old" workload slice
	// that will be preempted by the "new" workload slice, i.e., this annotates a "new" workload slice.
	// This annotation is alpha-level.
```

In addition, we aim to extend dynamic resizing support to RayCluster workloads. 
While the initial focus is on batch/v1.Job and RayCluster, the long-term objective is to generalize support for dynamically sized (elastic) workloads across other Kueue-integrated resource types.

## Motivation

Kueue currently lacks native support for resizing jobs. Any change in job size leads to the recreation of the associated Workload, resulting in job suspension and requeueing. 
This disrupts execution and hinders usability for elastic workloads like RayCluster, which rely on in-place autoscaling. 

To support such scenarios, Kueue must gracefully handle horizontal scale-up and scale-down operations without disrupting admitted jobs or re-acquiring quota.

### Goals

- Gracefully handle resize operations for Kueue-managed jobs by updating quota usage without suspending the job or terminating running pods.
  
### Non-Goals

- Vertical scaling of workloads – Kueue will only handle resize operations that scale Pods horizontally. 
- Support resize for other Kueue jobs such as Job, JobSet, etc (future).

> - Resizing of the Workload parent objects (`batchv1/Job`, RayJobs, etc.)

Reviewer: TBH I'm not sure what is meant in this sentence, the KEP aims to support resizing of parent objects. I see this was taken from the previous version of the KEP. However, I would suggest either clarify the intention or drop the point.

Author: My understanding,  it is not a goal of this proposal to orchestrate (initiate, coordinate, etc.) parent Job resizing.

Author: That said, I am ok removing this item altogether.

Reviewer: > My understanding, it is not a goal of this proposal to orchestrate (initiate, coordinate, etc.) parent Job resizing.

Oh I see, this makes sense. Maybe keep it then as: `Orchestration of resizing of the Job objects`.

Author: Updated, PTAL.

### Enablement

WorkloadSlices in Kueue are enabled through a combination of a Kubernetes feature gate and an opt-in annotation on individual Workload objects. 
At the cluster level, the WorkloadSlices feature must be enabled via the corresponding Kubernetes feature gate, which controls whether the controller logic for slicing is active. 

Once the feature gate is enabled, individual Workload objects can opt into slicing by including the `kueue.x-k8s.io/dynamically-sized-job: "true"` annotation. 
When both conditions are met, Kueue treats the Workload as eligible for partitioning into one or more WorkloadSlice objects, enabling fine-grained scheduling and 
execution across multiple clusters or within a single cluster. If the feature gate is disabled or the annotation is omitted (or set to "false"), 
the system defaults to the traditional single-Workload scheduling path for full backward compatibility.

#### Features
```golang
// owner: @ichekrygin
// kep: https://github.com/kubernetes-sigs/kueue/tree/main/keps/77-dynamically-sized-jobs
//
// WorkloadSlices enables workload-slices support.

> WorkloadSlices featuregate.Feature = "DynamicallySizedJob"

Author: Update (in the latest change).

> WorkloadSlices featuregate.Feature = "DynamicallySizedJob"

Reviewer: lgtm

```

#### WorkloadSliceAnnotation
```golang
// Workload slicing can be enabled for a specific integration job instance,
// provided that the integration supports workload slicing.
const (
  // EnabledAnnotationKey refers to the annotation key present on Job's that support
  // workload slicing.
  EnabledAnnotationKey = "kueue.x-k8s.io/dynamically-sized-job"
  // EnabledAnnotationValue refers to the annotation value. To enable
  // workload slicing for a given job, we match both annotation key and value.

>   EnabledAnnotationValue = "true"

Author: It is w/in a `const` block (not showing) here. 
https://github.com/kubernetes-sigs/kueue/pull/5510/files#diff-782ca8fda9b00ad0a088dac459ec6225ee53016ef8a64a82c33fb418aeea86bf

>   EnabledAnnotationValue = "true"

Reviewer: Ah, you are right actually indeed we define the consts in a couple of places in code: https://github.com/kubernetes-sigs/kueue/blob/cbe5ba6a6cb44160aae5a3b3e2b4bb461aac2ece/pkg/controller/jobs/pod/constants/constants.go#L30-L37. 

However:
1. the consts aren't really part of KEP: https://github.com/kubernetes-sigs/kueue/tree/main/keps/976-plain-pods
2. there is no even such practice in the core k8s, for example:
  - [rbac.authorization.kubernetes.io/autoupdate](https://github.com/kubernetes/kubernetes/blob/9e204ba03c2c95fd20dd18fcc6c40c3b16576592/pkg/apis/rbac/types.go#L40)
  - [safe-to-evict](https://github.com/kubernetes/autoscaler/blob/cc01c8756f900917021f7dda448db78290e572cf/cluster-autoscaler/utils/drain/drain.go#L146) in autoscaler
 
So I think it is totally ok to drop it from KEP and keep just implementation detail. We may or may not extract "true" as a constant then.

To address this, a new naming strategy will be introduced specifically for Workload slices, incorporating the owner object's generation value to ensure name uniqueness across revisions.
For example, `bachv1/Job`: 
```yaml
annotations:
  kueue.x-k8s.io/elastic-job: "true"
creationTimestamp: "2025-06-12T21:32:41Z"
generation: 1
labels:
  kueue.x-k8s.io/queue-name: demo
name: demo-slice
namespace: demo
resourceVersion: "190175"
uid: 78d1e61a-c399-4c09-800f-a0bbfb5fe71e
```

will result in a Workload name similar to: `job-demo-slice-75e0c`
Whereas, changing `generation: 2`: will result in a different name similar to: `job-demo-slice-e7080`

#### Scheduling and Preemption
Kueue’s scheduler and preemptor components will be augmented to support workload slice processing, specifically by enforcing preemption of the old workload slice regardless of queue capacity to make room for the new slice. 

Scheduling assignment is based on pod count capacity, accounting for the existing workload slice's allocated pods. 
For example, scaling a job up from `3` pods to `10` pods results in the creation of a new Workload representing all `10` pods; 

> however, the scheduling assignment will reflect only the `7` additional pods needed to admit the new slice, since `3` pods are already accounted for by the existing slice.

Reviewer: I'm not sure I understand that. I would assume the new Workload will also have all 10 Pods in the assignment. Since we "preempt" the old workload it will release the 3 units of quota.

Author: I understand, this part can definitely be a bit confusing.

The key detail is that **WorkloadSlice** preemption is _logical_, not _physical_. That is, the existing pods from the old slice are not evicted, they remain running and intact. As a result, while the new **Workload** includes all 10 pods in its spec, the scheduler only needs to assign quota for the additional 7 pods, the delta, since the first 3 will continue to run (i.e., will not be preempted).

Reviewer: Preemption and admission of the replacement workload is not an atomic operation. I think what should happen is:
1. the "old" slice is preempted and the quota of 3 is released
2. the "new" workload worth 10 is admitted (thus delta is 7)

The current wording suggests this is atomic operation. IIUC the "logical" quota increase is not reflected by any structures, either at the API level nor in memory. It is just "conceptual". Let me know if I'm missing something. 

If my mental model is correct I would suggest the following re-phrasing:

```
For example, scaling a job up from `3` pods to `10` pods results in the creation of a new Workload representing all `10` pods. Once the "new" Workload is at the queue head it preempts the "old" workload which releases quota
corresponding to 3 Pods. Then, the replacement Workload is admitted and books the quota corresponding to 
10 Pods. Note that in the process the quota usage was increased by 7 Pods.
```

Author: >Preemption and admission of the replacement workload is not an atomic operation

While preemption and admission are not strictly atomic operations, they occur within the same scheduling cycle, specifically within the same closure of workload processing ([scheduler.go#L213](https://github.com/kubernetes-sigs/kueue/blob/main/pkg/scheduler/scheduler.go#L213)). Barring transient errors, for every candidate workload, preemption is *immediately* followed by admission. We could think of this as **quasi-atomic** behavior from a scheduling perspective.

What’s important to emphasize is that while the *workload object* is preempted, the **pods** associated with the old workload are not. More importantly, those existing pods are effectively *absorbed* into the new workload.

In the example at hand, preempting an old workload with 3 running pods and admitting a new workload requesting 10 pods (7 of which are pending and schedule-gated), the new workload will effectively consist of 3 running pods and 7 un-gated pods. This transition avoids disruption while maintaining quota integrity.

>  IIUC the "logical" quota increase is not reflected by any structures, either at the API level nor in memory

That’s correct. While the quota increase is not explicitly reflected in any structure, either at the API level or in memory, it is implicitly represented by the new desired state: scaling from 3 to 10 pods.

A close analogy is how Kubernetes Deployments handle scaling. The updated Deployment declares a new desired replica count (e.g., from 3 to 10), rather than expressing the scale direction or magnitude (e.g., "+7"), which results in the creation of a new ReplicaSet. That ultimately yields 3 existing pods carried over, and 7 new pods created to fulfill the new desired state.

Reviewer: Thank you for the summary, I think when writing my last suggestion my understanding was fully inlined with the summary.

I find it acceptable as is, but pointing out as a place to potentially work on clarifying. And I'm aware the proposal is not perfect either. Maybe we could just emphasize a little more the increase is just eventual.

Author: Updated, capturing additional details. PTAL.

Reviewer: sgtm, thank you

Author: So... ahem, my previous comment wasn’t entirely accurate.

> We could think of this as quasi-atomic behavior from a scheduling perspective.

I overlooked the fact that during workload preemption, we *abandon* (or `continue`) scheduling the current workload and retry it in the next scheduling cycle. Just wanted to call that out here. That said, I still believe the behavior as described in the KEP remains accurate.

> A Workload that is preempted by a WorkloadSlice will be correctly marked with a Preempted status condition, including an appropriate reason and message.

Reviewer: > while only Finished is added by WorkloadSlicing.

Yes, but since "Finished" is added later it creates a small window when the preempted workload might be re-admitted. IIUC this could happen if you suddenly have enough quota to admit both the "old" workload and the "new" replacement workload at the same time. 

To prevent that from happening I see the possible approaches:
1. add `Finished` directly from Scheduler while preempting
2. tweak the reason for `Evicted` from `Preempted` to `DeactivatedDueToWorkloadSliceReplacement` - as deactivated workloads are not considered for re-admission

I'm totally ok with both approaches.

> A Workload that is preempted by a WorkloadSlice will be correctly marked with a Preempted status condition, including an appropriate reason and message.

Reviewer 2: I totally agree with Evicted with DeactivatedDueToWorkloadSliceReplacement instead of Preempted.

My additional question is what if they enable objectRetentionPolicy (https://kueue.sigs.k8s.io/docs/tasks/manage/setup_object_retention_policy/#set-up-a-retention-policy)?

Which condition lastTransitionTime (Evicted vs Finished) is respected?

> A Workload that is preempted by a WorkloadSlice will be correctly marked with a Preempted status condition, including an appropriate reason and message.

Author: done.

> A Workload that is preempted by a WorkloadSlice will be correctly marked with a Preempted status condition, including an appropriate reason and message.

Author: Thinking about this a bit more:

```yaml
- type: Preempted
  reason: WorkloadSliceReplacement
- type: Evicted
  reason: DeactivatedDueToWorkloadSliceReplacement
- type: Finished  # added later
  reason: WorkloadSliceReplacement
```

In the context of replacing an old workload slice with a new one, does this really qualify as **preemption**?

To me, it feels more like an **eviction without preemption**. Preemption generally implies that the workload might eventually be re-admitted, as @mimowo pointed out:

> Preempted workload might be re-admitted

But in this case, the intent is to evict and **replace** the old workload slice, not pause or defer it for later. That seems like a different semantic.

It reminds me of the distinction in Kubernetes between a Pod evicted by the scheduler (and still eligible for rescheduling) versus a Pod evicted by the kubelet and marked as `Failed`, with no opportunity for rescheduling. For workload slices, we’re closer to the latter case.

So maybe we could simplify this by marking the old slice as **evicted and finished**, without using `Preempted` at all. What do you think?

> A Workload that is preempted by a WorkloadSlice will be correctly marked with a Preempted status condition, including an appropriate reason and message.

Reviewer: Except for Pod integration, in Kueue the unit of preemption (and eviction) is Workload (like Job). When Workload is preempted the Job gets suspended, and all Pods deleted. Kueue may already evict a running, healthy workload, if it is low-priority (for example).

> A Workload that is preempted by a WorkloadSlice will be correctly marked with a Preempted status condition, including an appropriate reason and message.

Author: > In summary, I'm ok with either:
> 
> 1. `Finished` only - minimalist API approach
> 2. `Finished + Evicted(Deactivated) + Preempted` - conditions which are added naturally by reusing the code paths for Preemption
> 
> Options when `Evicted` is set, but not `Preempted` feel a bit artificial to me. Let me know if I'm missing something.

I lean toward the first option, mainly for the sake of simplicity, both in how we apply status conditions and, more importantly, in how we check them.

Having three separate conditions to reflect essentially the same state feels unnecessarily complex and risks introducing ambiguity. For example, it could lead to inconsistent or unexpected combinations, like `Preempted` without `Evicted`, or vice versa.

Specifically regarding Kueue’s current usage of `Preempted` and `Evicted`, from what I can tell, they are always applied together. I might be missing something, but I haven’t come across any cases where one is set without the other. If that’s correct, maybe we could even consolidate those into a single condition, possibly just `Evicted`, but that is outside of this KEP context. 

In the case of workload slice aggregation, here I am using the term “aggregation” deliberately to distinguish this from traditional preemption or eviction. It’s really neither in the strict sense—we’re just replacing one workload with another. I did consider introducing a new `Aggregated` condition, but decided against it to avoid introducing yet another condition with ripple effects across the codebase.

So, if setting a single `Finished` condition with a reason like `"WorkloadSliceAggregation"` covers the intent and already supported but the current Kueue's workload lifecycle handling, I’d definitely vote for that!

> A Workload that is preempted by a WorkloadSlice will be correctly marked with a Preempted status condition, including an appropriate reason and message.

Reviewer: sgtm, Im fine with 1. The only reason for 3 I was thinking about is that it probably saves a single 'if' as by default these would be set
Setting the conditikns is also coupled with metrics updates, however, we probably need to revisit metrics for the feature anyway.

This approach avoids unpredictable flavor switching but may lead to the new slice remaining in a Pending state if the original flavor lacks sufficient resources.

Future enhancements may introduce support for controlled flavor reassignment or user-defined policies to express preferences or tolerances for such transitions.

#### Garbage Collection of Preempted Workload Slices
Currently, a Workload can become outdated, marked by Kueue as "out-of-sync", and subsequently transitioned to the "Finished" state.
This also applies to outdated Workload slices: when multiple successive updates are issued for a given Job, any superseded slices are marked as "Finished" to reflect their obsolescence.

Preempted (i.e., aggregated) Workload slices are marked as Finished.
Under the current design proposal, all finished workload slices are retained indefinitely and are not garbage collected.
To address potential resource buildup, a `PreemptedWorkloadSliceHistory` mechanism whether as a configuration option or an API field on the Workload—could be introduced to limit the number of retained inactive slices, 

> similar to how `revisionHistoryLimit` is used in Kubernetes Deployments to manage ReplicaSet history.

Reviewer: Let's defer this to Beta (add the graduation point). For alpha, if we aim 0.13 we will not have time for it. Also, I think ObjectRetentionPolicy should handle it in most cases.

Author: Agreed, it is already in "Beta" section.

Author: For posterity, as [currently defined](https://kueue.sigs.k8s.io/docs/tasks/manage/setup_object_retention_policy/#set-up-a-retention-policy):

```yaml
objectRetentionPolicies:
  workloads:
    afterFinished: "1h"
    afterDeactivatedByKueue: "1h"
```

the `objectRetentionPolicy` is exclusively duration-driven, which may not be the best fit for garbage collecting *aggregated* workload slices.

As you suggested, adding a `finishedWorkloadLimit` (or similar) field to the policy definition would be preferable.

that represents a partition of the original workload. Each slice contains a `PodSet.Count` identical to the Job’s definition, along with scheduling metadata needed for handling preemption of the old workload (slice).

At most two active workload slices can exist for any single Job in Kueue: the current admitted slice and a newly created slice during a scaling operation. 
This ensures a controlled transition between slices while maintaining consistency and avoiding resource contention and race conditions.

To facilitate precise resource scheduling and effective workload preemption, every new job configured with slice enablement will be appropriately annotated using `EnabledAnnotationKey/Value` mentioned above.
Furthermore, each individual workload slice will record the identity of any preempting slice it encounters.

```golang

const (
	// WorkloadPreemptibleSliceNameKey is the annotation key used to capture an "old" workload slice
	// that will be preempted by the "new" workload slice, i.e., this annotates a "new" workload slice.

> 	WorkloadPreemptibleSliceNameKey = "kueue.x-k8s.io/workload-preemptible-slice"

Author: It’s similar, but ultimately serves a different purpose.

The annotation `kueue.x-k8s.io/dynamically-sized-job: "true"` is used to indicate that a given Job is enabled for the dynamic resize feature (in addition to requiring the feature flag to be set).

In contrast, the `kueue.x-k8s.io/workload-preemptible-slice` annotation is applied to a new workload instance (i.e., a slice) and refers to the name of the existing workload (the “old slice”) that it is intended to preempt.

> 	WorkloadPreemptibleSliceNameKey = "kueue.x-k8s.io/workload-preemptible-slice"

Author: Foreshadowing future **multi-Kueue** support, the `kueue.x-k8s.io/workload-preempting-slice` annotation will be set on the old workload to indicate the name of the new slice that is preempting it.

> 	WorkloadPreemptibleSliceNameKey = "kueue.x-k8s.io/workload-preemptible-slice"

Reviewer: sgtm, thanks for explaining

)
```

##### Naming
Currently, Kueue uses a deterministic Workload naming strategy by generating consistent names for the same Job instance via `GetWorkloadNameForOwnerWithGVK`.
As a result, Workload names are not unique across different revisions of the same owner object (e.g., a Job), which makes this approach unsuitable for representing Workload slices.

To address this, a new naming strategy will be introduced specifically for Workload slices, incorporating the owner object's generation value to ensure name uniqueness across revisions.
For example, `bachv1/Job`: 
```yaml
annotations:
  kueue.x-k8s.io/dynamically-sized-job: "true"
creationTimestamp: "2025-06-12T21:32:41Z"
generation: 1
labels:
  kueue.x-k8s.io/queue-name: demo
name: demo-slice
namespace: demo
resourceVersion: "190175"
uid: 78d1e61a-c399-4c09-800f-a0bbfb5fe71e
```

will result in a Workload name similar to: `job-demo-slice-75e0c`

> Whereas, changing `generation: 2`: will result in a different name similar to: `job-demo-slice-e7080`

Author: Excellent question, @mimowo. 

The `GetWorkloadNameForOwnerWithGVK` function is invoked only within the `ConstructWorkload` context. Therefore, if there are any valid changes to the **Job** object that result in a new revision, but do not modify the pod count (e.g., replicas, parallelism, etc.) in a way that increases it, those changes will effectively be ignored, as they do not result in a new workload (slice) being created.

#### Scheduling and Preemption
Kueue’s scheduler and preemptor components will be augmented to support workload slice processing, specifically by enforcing preemption of the old workload slice regardless of queue capacity to make room for the new slice. 

Scheduling assignment is based on pod count capacity, accounting for the existing workload slice's allocated pods. 
For example, scaling a job up from `3` pods to `10` pods results in the creation of a new Workload representing all `10` pods; 
however, the scheduling assignment will reflect only the `7` additional pods needed to admit the new slice, since `3` pods are already accounted for by the existing slice.

A Workload that is preempted by a WorkloadSlice will be correctly marked with a Preempted status condition, including an appropriate reason and message.
```golang
// WorkloadSlicePreemptionReason indicates the Workload was preempted due to
// the workload slice succession (roll-up/aggregation).
WorkloadSlicePreemptionReason string = "WorkloadSlicePreemption"
...
// WorkloadEvictedByWorkloadSliceAggregation indicates that the workload was
// deactivated as a result of workload slice aggregation.

> WorkloadEvictedByWorkloadSliceAggregation = "WorkloadSliceAggregation"	

Author: @mimowo, thank you for clarification - removing.

Flavor selection is influenced by factors such as quota availability, taint tolerations, node label selectors, and, where applicable, topology-aware scheduling policies.

When scaling up a previously admitted workload slice, the current implementation enforces a sticky flavor constraint, meaning that the new slice must reuse the originally assigned flavor, even if other eligible flavors have available capacity. 
This approach avoids unpredictable flavor switching but may lead to the new slice remaining in a Pending state if the original flavor lacks sufficient resources.

Future enhancements may introduce support for controlled flavor reassignment or user-defined policies to express preferences or tolerances for such transitions.

#### Garbage Collection of Preempted Workload Slices
Currently, a Workload can become outdated, marked by Kueue as "out-of-sync", and subsequently transitioned to the "Finished" state.
This also applies to outdated Workload slices: when multiple successive updates are issued for a given Job, any superseded slices are marked as "Finished" to reflect their obsolescence.

> Preempted (i.e., aggregated) Workload slices are first marked for Deactivation and then transitioned to the Finished state.

Author: It was my misunderstanding of Kueue’s workload deactivation process prior to marking it as finished.
If deactivation is not required, we can skip that phase and transition the workload directly to _Finished_.

> Preempted (i.e., aggregated) Workload slices are first marked for Deactivation and then transitioned to the Finished state.

Reviewer: sgtm

> Under the current design proposal, all deactivated slices are retained indefinitely and are not garbage collected.

Author: We could, although the Kubernetes `deployment.spec.revisionHistoryLimit` model might be a better fit for this use case. WDYT?

> Under the current design proposal, all deactivated slices are retained indefinitely and are not garbage collected.

Reviewer: We need a mechanism that will be JobCRD independent. Since we already have the ObjectRetentionPolicy Im sceptical if we need more code for GC. At least for Alpha it does not seem required. Also the two mechanisms could race sometimes.

Maybe going forward we consider an extension to ObjectRetentionPolicy to specify FinishedWorkloadLimit, wdyt?

## Design Details

To support horizontal scaling of Jobs, we will introduce the concept of a "WorkloadSlice". A WorkloadSlice is a Workload object with an owner reference to the original Workload for a Job. 
Workload Slices represent per-replica changes to a Job that were not initially accounted for when the Job was created. **It's an internal state, not an extra CRD.**

The benefit of Workload Slices is that Kueue can evaluate admission on a per-replica basis without changing the existing semantics of the Workload API. 
Once a WorkloadSlice is admitted, it will replace the original/old WorkloadSlice. 
The latter will be deactivated and marked as finished.
- Each WorkloadSlice inherits the resource flavor assignment of the original admitted slice, enforcing flavor consistency across scale-ups. 
- In MultiKueue, Workload Slices would go into both clusters (management and workload) in a multi-cluster environment.
- Workload Slices will be created by Kueue and use identical PodTemplate.
- Workload Slices will belong to the same resource flavor as the top-level Workload that was initially admitted (for more details see "Resource Flavors" subsection)

### Enablement

> WorkloadSlices in Kueue are enabled through a combination of a Kubernetes feature gate and an opt-in annotation on individual Workload objects. 

Reviewer 2: https://kueue.sigs.k8s.io/docs/installation/#change-the-feature-gates-configuration is Kueue feature gate.
https://kubernetes.io/docs/reference/command-line-tools-reference/feature-gates/ is Kubernetes feature gate.

I'm ok with either approach (dropping "Kubernetes" term vs replacing "Kubernetes" with "Kueue").

> WorkloadSlices in Kueue are enabled through a combination of a Kubernetes feature gate and an opt-in annotation on individual Workload objects. 

Author: If there is a consensus (as it appears), I would leave it as "Kueue feature gate".

> WorkloadSlices in Kueue are enabled through a combination of a Kubernetes feature gate and an opt-in annotation on individual Workload objects. 

Author: updated.

> WorkloadSlices in Kueue are enabled through a combination of a Kubernetes feature gate and an opt-in annotation on individual Workload objects. 

Reviewer 2: LGTM

> At the cluster level, the WorkloadSlices feature must be enabled via the corresponding Kubernetes feature gate, which controls whether the controller logic for slicing is active. 

Author: Relate to: https://github.com/kubernetes-sigs/kueue/pull/5648#discussion_r2201442433

Once the feature gate is enabled, individual Workload objects can opt into slicing by including the `kueue.x-k8s.io/elastic-job: "true"` annotation. 
When both conditions are met, Kueue treats the Workload as eligible for partitioning into one or more WorkloadSlice objects, enabling fine-grained scheduling and 
execution across multiple clusters or within a single cluster. If the feature gate is disabled or the annotation is omitted (or set to "false"), 
the system defaults to the traditional single-Workload scheduling path for full backward compatibility.

#### Features
```golang
// owner: @ichekrygin
// kep: https://github.com/kubernetes-sigs/kueue/tree/main/keps/77-dynamically-sized-jobs
//
// ElasticJobsViaWorkloadSlices enables support for horizontal-scale in jobs.
ElasticJobsViaWorkloadSlices featuregate.Feature = "ElasticJobsViaWorkloadSlices"

> ```

Reviewer 2: LGTM

#### WorkloadSliceAnnotation
```golang
// Workload slicing can be enabled for a specific integration job instance,
// provided that the integration supports workload slicing.
const (
  // EnabledAnnotationKey refers to the annotation key present on Job's that support
  // workload slicing.
  // This annotation is alpha-level.
  EnabledAnnotationKey = "kueue.x-k8s.io/elastic-job"
  // EnabledAnnotationValue refers to the annotation value. To enable
  // workload slicing for a given job, we match both annotation key and value.

>   EnabledAnnotationValue = "true"

Reviewer 2: If this is just for backward compatibility, we probably can abandon this annotation when we promote the Workload API version to v1beta2.

But, that is not now.

Reviewer 2: > the mechanism is based on scheduling gates, and this would mean performance hit for users not using ElasticJobs. I think removing the scheduling gate from every Pod can be felt by some large-scale users. Sure, for most users it will not be an issue, so we may have "AssumeAllJobsAreElastic" in global configuration, but I would defer it for later.

Actually, as my initial thought was moving this annotation functionality to Global Config (AssumeAllJobsAreElastic).
However, I wanted to think more if the config is useful. In other words, for now, this should be annotation. But, in the future versioin, it might not be annotation.

that represents a partition of the original workload. Each slice contains a `PodSet.Count` identical to the Job’s definition, along with scheduling metadata needed for handling preemption of the old workload (slice).

At most two active workload slices can exist for any single Job in Kueue: the current admitted slice and a newly created slice during a scaling operation. 
This ensures a controlled transition between slices while maintaining consistency and avoiding resource contention and race conditions.

To facilitate precise resource scheduling and effective workload preemption, every new job configured with slice enablement will be appropriately annotated using `EnabledAnnotationKey/Value` mentioned above.
Furthermore, each individual workload slice will record the identity of any preempting slice it encounters.

```golang

const (
	// WorkloadPreemptibleSliceNameKey is the annotation key used to capture an "old" workload slice
	// that will be preempted by the "new" workload slice, i.e., this annotates a "new" workload slice.
	// This annotation is alpha-level.

> 	WorkloadPreemptibleSliceNameKey = "kueue.x-k8s.io/workload-preemptible-slice"

Reviewer: I think this annotation isn't really necessary. We can effectively use the Fieldindexer as suggested in the impl PR: https://github.com/kubernetes-sigs/kueue/pull/5510#discussion_r2198243591

Reviewer: Ok, as discussed in the related thread, I'm ok to keep it for optimizing performance. However, I prefer a more self-explanatory name like `kueue.x-k8s.io/workload-slice-replacement-for`

Reviewer 2: I think this could be moved to status field.
However, let's postpone the API expansion decision later, then use this annotation for now.
@ichekrygin Could you replace the annotation name as @mimowo mentioned?

Author: > I think this could be moved to status field.

To be honest, neither the annotation nor, even less so, the status field are ideal places to define desired state. If we follow the Kubernetes paradigm strictly, the only appropriate place to express desired state is the object’s `spec`.

Specifically, the reason why the status is not suitable is that we should be able to wipe the entire status clean and still be able to reconcile the object, in this case, the Workload, based solely on its spec to reach the desired state. The annotation is a bit of a "gray-area" in this regard. I understand that Kueue currently relies on the status subresource as a kind of “quasi-spec,” and that may be acceptable as an interim solution.

That said, I agree that we should revisit this during an API review and possibly consider a refactor for better alignment with core Kubernetes conventions.

Reviewer: > That said, if the use of annotations becomes a blocker,

It is not a blocker to me while in alpha, but rename it to `kueue.x-k8s.io/workload-slice-replacement-for`, and add a point in graduation criteria for Beta to revisit this.

Author: Updated, PTAL and advice if this can be resolved (or resolve).

)
```

> ##### Naming

Reviewer 2: When there are old and new Workload (slice) in the cluster, how to identify the latest one from batch/v1 Job object?

Reviewer 2: I see. SGTM.
Could you mention that we always respect the latest created WorkloadSlice?

Currently, Kueue uses a deterministic Workload naming strategy by generating consistent names for the same Job instance via `GetWorkloadNameForOwnerWithGVK`.
As a result, Workload names are not unique across different revisions of the same owner object (e.g., a Job), which makes this approach unsuitable for representing Workload slices.

To address this, a new naming strategy will be introduced specifically for Workload slices, incorporating the owner object's generation value to ensure name uniqueness across revisions.
For example, `bachv1/Job`: 
```yaml
annotations:
  kueue.x-k8s.io/elastic-job: "true"
creationTimestamp: "2025-06-12T21:32:41Z"
generation: 1
labels:
  kueue.x-k8s.io/queue-name: demo
name: demo-slice
namespace: demo
resourceVersion: "190175"
uid: 78d1e61a-c399-4c09-800f-a0bbfb5fe71e
```

will result in a Workload name similar to: `job-demo-slice-75e0c`

> Whereas, changing `generation: 2`: will result in a different name similar to: `job-demo-slice-e7080`

Author: I may not fully understand the request. The algorithm id identical to the current implementation, and includes just one extra field (generation) during "hash suffix" value calculation.

> Whereas, changing `generation: 2`: will result in a different name similar to: `job-demo-slice-e7080`

Reviewer 2: Sorry for your confusion. I want to clarify what is format to compile this hash (75e0c).
The current form is `$PREFIX_NAME-$HASH` where $HASH is compiled by group, kind, owner name, and owner UID.
Will this proposal try to add new element, generation to $HASH?
If yes, I want to mention that here.

> Whereas, changing `generation: 2`: will result in a different name similar to: `job-demo-slice-e7080`

Author: > For example, we could say that the current Wokrload name suffix ...

@tenzen-y , I believe the KEP already says this in more or less the same words under the "Naming" section. PTAL and let me know if you'd like me to update or clarify anything there.

> #### Scheduling and Preemption

Reviewer 2: What if the flavor assigner assigns a different flavor to a new Workload from an old Workload?

Will you implement any mechanism to assign the same flavor to a new Workload as an old Workload?
Will the flavor assigner filter out the different flavors from an old Workload?

Author: This is actually captured in the **Design Details** section:

> *Each WorkloadSlice inherits the resource flavor assignment of the original admitted slice, enforcing flavor consistency across scale-ups.*

> Will you implement any mechanism to assign the same flavor to a new Workload as an old Workload?

Yes, that’s already a standing requirement.

> Will the flavor assigner filter out the different flavors from an old Workload?

I’m not sure I fully understand the question, could you clarify what you mean by “filter out”?

Author: > The assigner just retries the same flavor with increased count, right?

Yes, that is correct.

   4. The newly admitted Pods begin scheduling and running, resulting in the Job operating with the increased parallelism level.

#### Story 3 – batch/v1.Job in Multi-Cluster Configuration

This story mirrors the behaviors in Stories 1 and 2 but applies them within a multi-cluster (MultiKueue) setup, 
where WorkloadSlices are propagated, admitted, and scheduled across multiple clusters.

### Notes/Constraints/Caveats (Optional)

If Kueue needs to preempt a resized Job, it will preempt the entire Job as a single unit, regardless of whether the Job has undergone a scale-up operation.

## Design Details

To support horizontal scaling of Jobs, we will introduce the concept of a "WorkloadSlice". A WorkloadSlice is a Workload object with an owner reference to the original Workload for a Job. 
Workload Slices represent per-replica changes to a Job that were not initially accounted for when the Job was created. **It's an internal state, not an extra CRD.**

The benefit of Workload Slices is that Kueue can evaluate admission on a per-replica basis without changing the existing semantics of the Workload API. 
Once a WorkloadSlice is admitted, it will replace the original/old WorkloadSlice. 

> The latter will be deactivated and marked as finished.

Reviewer: ```suggestion
The replaced slice will be marked as finished.
```
nit, as we discussed on the implementation PR, the deactivation is not needed

##### Preemption Disambiguation.

Here, we’re using the term “preemption” to describe the process of replacing an old workload slice with a new one. 
At the same time, Kueue already uses the concept of workload preemption to mean evicting a workload and deferring its scheduling until capacity becomes available.

In the context of workload slices, it may be more helpful to think of this replacement process as **aggregation** rather than preemption. 
Aggregation involves terminating the old slice and activating the new one. In the strict sense, neither eviction nor preemption (as traditionally defined) actually occurs during this process.

As such, we mark the old slice with the `Finished` condition. This ensures it is removed from the in-memory ClusterQueue snapshot, and its resource usage is released and attributed to the new slice that replaces it.

A Workload that is preempted by a WorkloadSlice will be correctly marked with a Finished status condition, including an appropriate reason and message.
```yaml
- type: Finished
  status: True

>   reason: WorkloadSliceAggregation

Reviewer 2: +1 on WorkloadSliceReplacement.

>   reason: WorkloadSliceAggregation

Author: Updated.

### Phase 1 - batchv1/Job WorkloadSlices Support in Single-Cluster Configuration.

Scaling up and down for `batch/v1.Job` will be the initial phase of the MVP, as it builds on stable and well-understood Kubernetes components without requiring additional integration work.

#### Scale Down

1. Create a Job with workload-slice enablement and a parallelism value greater than 1. 
2. Observe that a corresponding Workload is created and admitted, and that the Job's pods are in the Running state, matching the specified parallelism. 
3. Update the Job by reducing its parallelism value. 
4. Confirm that the existing Workload is updated accordingly, with WorkloadSpec.PodSets[0].Count reflecting the new parallelism value. 
5. Observe that the number of running pods is reduced, while the remaining pods continue to run uninterrupted.

#### Scale Up

1. Create a `batch/v1.Job` with workload-slice enablement and an initial `parallelism` value greater than 0.
2. Verify that a corresponding `Workload` is created and admitted, and that the Job’s pods are in the `Running` state, matching the specified parallelism.
3. Increase the Job’s `parallelism` value to trigger a scale-up event.

> 4. Confirm that a new `Workload` slice is created and admitted, reflecting the increased pod count, while the previous slice is deactivated and marked as `Finished`.

Reviewer: ```suggestion
4. Confirm that a new `Workload` slice is created and admitted, reflecting the increased pod count, while the previous slice is marked as `Finished`.
```

## Summary

Enable dynamic resizing of Kueue-managed batch/v1.Job workloads by supporting in-place horizontal scaling of job parallelism. 
The primary goal is to allow jobs to scale their pod count without requiring suspension and requeueing, improving scheduling efficiency and reducing disruption during resource fluctuations.

> While the initial focus is on batchv1/Job type, the long-term goal is to generalize support for dynamically sized (elastic) workloads across other Kueue-integrated resource types like RayClusters.

Reviewer 3: If possible, I suggest we focus on both batchv1/Job and RayCluster, because it is very common to enable autoscaling for RayCluster which was the initial motivation for this proposal. I'm happy to help review, test and implement anything specific to RayCluster if needed to make this happen.

Author: Updated, PTAL. 

Please review the remaining KEP, specifically "Phases for MVP," and suggest RayClusters MVP targets if needed.

While the initial focus is on batch/v1.Job and RayCluster, the long-term objective is to generalize support for dynamically sized (elastic) workloads across other Kueue-integrated resource types.

## Motivation

Kueue currently lacks native support for resizing jobs. Any change in job size leads to the recreation of the associated Workload, resulting in job suspension and requeueing. 
This disrupts execution and hinders usability for elastic workloads like RayCluster, which rely on in-place autoscaling. 

To support such scenarios, Kueue must gracefully handle horizontal scale-up and scale-down operations without disrupting admitted jobs or re-acquiring quota.

### Goals

- Gracefully handle resize operations for Kueue-managed jobs by updating quota usage without suspending the job or terminating running pods.
  - Support dynamic updates to `batch/v1.Job` parallelism, allowing changes to pod count without deleting existing pods.
  
### Non-Goals

- Vertical scaling of workloads – Kueue will only handle resize operations that scale Pods horizontally. 
- Support resize for other Kueue jobs such as Job, JobSet, etc (future).

> - Resizing of the Workload parent objects (`batchv1/Job`, RayJobs, etc.)

Author: @weizhaowz, @andrewsykim  - can we resolve this conversation?

  - Dynamically adjust quotas in ClusterQueue based on scaling events.

For the MVP, dynamic resizing will be available through an opt-in mechanism for selected, supported Kueue frameworks—starting with batch/v1.Job and, optionally, RayCluster workloads with autoscaling enabled. 
Additionally, WorkloadSlice will be supported in the multiKueue context as part of the MVP scope.

### User Stories

#### Story 1 - batchv1/Job scale-down

1. The user creates a `batch/v1.Job` with workload-slice enablement explicitly opted in.
2. Kueue admits the Job based on its requested resources, creating the corresponding Workload and Job's Pods. The Pods begin running as usual.
3. The user updates the Job by reducing its parallelism, triggering a scale-down event:
   1. The associated Workload is updated to reflect the new, lower pod count in its spec. 
   2. The local queue’s capacity is updated accordingly to reflect the reduced resource usage. 
   3. One running Job pod is terminated, while the remaining Pods continue running uninterrupted.

#### Story 2 - batchv1/Job scale-up
1. The user creates a `batch/v1.Job` with workload-slice enablement explicitly opted in.

> 2. Kueue admits the Job based on its requested resources, creating the corresponding Workload and Job’s Pods. The Pods begin running as usual.

Author: > 1. the parent workload is created but not admitted yet, and the owner decided to increase parallelism, so workslices are created as well, is it possible for the the workslice get admitted before the ownerpreference workload(I guess not, because if they are with the same priority and the parent workload is in front of the workslice in the local queue)?

For workloads that are not yet admitted, specifically, those without a quota reservation, we update the existing Workload with the new requirements during a scale-up event. This aligns with current Kueue behavior (see [reconciler.go#L814](https://github.com/kubernetes-sigs/kueue/blob/main/pkg/controller/jobframework/reconciler.go#L814)). In other words, we do not create a new WorkloadSlice for non-admitted workloads during scale-up.

Regarding edge cases or race conditions,
In theory, it’s possible for the job reconciler (ReconcileGenericJob) to operate on a stale version of the Workload, for instance, just after the scheduler has admitted it but before the reconciler has observed the update. In that scenario, the reconciler's attempt to update the outdated Workload will result in a conflict. This will trigger a requeue and a subsequent reconcile, at which point it will observe the current admitted state and proceed to scale the workload accordingly, i.e., by creating one or more WorkloadSlices.

> 2. Kueue admits the Job based on its requested resources, creating the corresponding Workload and Job’s Pods. The Pods begin running as usual.

Author: > 2. is it possible that the parent workload completes before the workslice is admitted? If so, will the workslices be marked as "Finished"?

I suspect it is not likely, and in practice, it should not happen (frequently?).

A `WorkloadSlice` is created as a response to a scale-up event. In the case of dynamic resizing for `batch/v1.Job`, all newly created pods are **schedule-gated** until the corresponding `WorkloadSlice` is both created and admitted. This means the job cannot progress toward completion based on scaled-up replicas unless the slice is admitted. Therefore, if the job completes, it typically indicates one of two things:

* The job met its completion criteria using the currently admitted workload(s), or
* The job was deleted (manually or otherwise).

In the deletion case, all associated `Workload` and `WorkloadSlice` objects are garbage collected automatically,
In the completion case, Kueue marks the current active `Workload` (which may be the original or a slice) as `Finished`.

Now, if I understand your question correctly, you're asking whether a race is possible where the parent job completes *while* we're in the middle of a scale-up, potentially leading to two active `Workload` objects:

1. The original `Workload`, which may still be undergoing preemption to make room for,
2. A newly created `WorkloadSlice` with the increased parallelism, created just before the job completes.

In this rare case, yes, we may briefly observe both `Workload` objects active. However, the job reconciler (`ReconcileGenericJob`) only processes **one** `Workload` per job at a time. It should:

* First, update the new `WorkloadSlice` to mark it as `Finished`, since the job has already completed, and then
* On a subsequent reconcile, update the original `Workload` (if still present and not already preempted), marking it as `Finished` as well.

This logic ensures consistency and clean termination, even across boundary cases like this.

> 2. Kueue admits the Job based on its requested resources, creating the corresponding Workload and Job’s Pods. The Pods begin running as usual.

Author: Excellent observations and insightful questions, @weizhaowz, please keep them coming!

3. Update the Job by reducing its parallelism value. 
4. Confirm that the existing Workload is updated accordingly, with WorkloadSpec.PodSets[0].Count reflecting the new parallelism value. 
5. Observe that the number of running pods is reduced, while the remaining pods continue to run uninterrupted.

#### Scale Up

1. Create a `batch/v1.Job` with workload-slice enablement and an initial `parallelism` value greater than 0.
2. Verify that a corresponding `Workload` is created and admitted, and that the Job’s pods are in the `Running` state, matching the specified parallelism.
3. Increase the Job’s `parallelism` value to trigger a scale-up event.
4. Confirm that a new `Workload` slice is created and admitted, reflecting the increased pod count, while the previous slice is deactivated and marked as `Finished`.
5. Observe that the total number of running pods increases to match the updated parallelism, with the original pods continuing to run without disruption.

### Phase 2 – RayCluster WorkloadSlice Support in Single-Cluster Configuration

> This phase mirrors the steps from Phase 1 but is adapted specifically for RayCluster workloads, taking into account their autoscaling behavior and internal lifecycle management.

Reviewer 3: I think focusing on autoscaling only is fine. I suspect the underlying mechanism to handle autoscaling or manual scaling will be the same anyways

[deprecation-policy]: https://kubernetes.io/docs/reference/using-api/deprecation-policy/
-->

Here’s a structured and detailed **Graduation Criteria** section for KEP-77: *Dynamically Sized Jobs*, tailored to your current implementation of `WorkloadSlice` and phased MVP plan.

---

### Graduation Criteria

#### Alpha

* [x] Feature is gated by the `ElasticJobsViaWorkloadSlices` feature gate.
* [x] WorkloadSlice is implemented as an internal Kueue construct (no new CRDs).
* [x] Scale-up events result in the creation of new `WorkloadSlice`s; scale-downs result in updates to the original `Workload`.
* [x] Pod-level gating using `PodSchedulingGates` is integrated with slice admission.
* [x] Sticky flavor behavior is enforced: scale-ups must reuse the original assigned flavor.

> * [x] Superseded slices are deactivated and marked as `Finished` once the new slice is admitted.

Reviewer: ```suggestion
* [x] Superseded slices are marked as `Finished` once the new slice is admitted.
```

* [ ] Formal conformance tests validate end-to-end behavior for:
    * Horizontal scale-up and scale-down
    * Slice transitions (preemption/deactivation)
    * Sticky flavor enforcement
    * Multi-cluster propagation
* [ ] At least one additional framework beyond `batch/v1.Job` (e.g., RayCluster) integrates and validates the WorkloadSlice flow.
* [ ] Slice lifecycle events (e.g., admitted, preempted, finished) are observable via `kubectl describe workload` or equivalent API tools.
* [ ] Slice preemption is consistently handled and visible in workload status conditions.
* [ ] All Kueue core controllers (scheduler, preemptor, queue manager) are validated under slice-enabled workloads.
* [ ] Dynamic resizing is enabled for all Kueue-managed workloads that support the elastic-job feature (including JobSet, RayJob, Kubeflow jobs, etc.)
* [ ] Re-evaluate the WorkloadSlice implementation to ensure compatibility with elastic workloads, considering all current and emerging alternatives within Kueue.
* [ ] Re-evaluate currently disallowed per-job-instance combination of enabled PartialAdmission and ElasticJobs.

> 

Author: Added.

| Terminating Pods         | ❌ No              | Identified by a `deletionTimestamp`  |
| Mirror Pods              | ❌ No              | Created directly by the kubelet      |
| Pods outside quota scope | ❌ No              | If the `ResourceQuota` uses `scopes` |

---

### PodScheduling Readiness and ResourceQuota

**PodScheduling Readiness** is a mechanism that ensures a pod is only eligible for scheduling when it meets specific conditions. This is especially useful for multi-stage scheduling, resource reservation, or coordination with external systems.

An important observation: by default, pods that are gated using `PodSchedulingGates` are still counted toward the namespace's `ResourceQuota` if they do not match any of the exclusion conditions listed above.

As a result, when `PodSchedulingGates` are used in a namespace with quota enforcement, it is possible to exhaust the available quota with ScheduleGated pods that are not yet schedulable. This can block additional pods from being admitted, even though none are actively running.

## Alternatives

- Require users to manage resizing manually by recreating jobs.

> - Defer support for elastic workloads to higher-level controllers (e.g., RayOperator), leaving Kueue unaware of scale operations.

Author: @mimowo, thank you for the suggestion regarding resizeRequest. Do you have a GitHub issue (or any other link) I could reference to point others to this alternative?

> - Defer support for elastic workloads to higher-level controllers (e.g., RayOperator), leaving Kueue unaware of scale operations.

Reviewer: Opened, ptal: https://github.com/kubernetes-sigs/kueue/issues/5897. 

This is not a blocker to the effort by any means. Having a prototype for both options would be great for decision making in long run.
